src/extractor/guidance/mergable_road_detector.cpp

```diff
@@ -0,0 +1,427 @@
+     * effect. We therefore only prune away the beginning, if the roads don't meet up again as well.
+     */
+    if (!connect_again)
+    {
+        prune(coordinates_to_the_left);
+        prune(coordinates_to_the_right);
+    }
+
+    const auto are_parallel = util::coordinate_calculation::areParallel(coordinates_to_the_left,
+                                                                        coordinates_to_the_right);
+
+    if (!are_parallel)
+        return false;
+
+    // compare reference distance:
+    const auto distance_between_roads = util::coordinate_calculation::findClosestDistance(
+        coordinates_to_the_left[coordinates_to_the_left.size() / 2], coordinates_to_the_right);
+
+    const auto combined_road_width =
+        0.5 *
+        (std::max<int>(
+             1, node_based_graph.GetEdgeData(lhs.eid).road_classification.GetNumberOfLanes()) +
+         std::max<int>(
+             1, node_based_graph.GetEdgeData(rhs.eid).road_classification.GetNumberOfLanes())) *
+        ASSUMED_LANE_WIDTH;
+    return distance_between_roads <= combined_road_width + 8;
+}
+
+bool MergableRoadDetector::ConnectAgain(const NodeID intersection_node,
+                                        const MergableRoadData &lhs,
+                                        const MergableRoadData &rhs) const
+{
+    /* compute the set of all intersection_nodes along the way of an edge, until it reaches a
+     * location with the same name repeatet at least three times
+     */
+    const auto left_connection =
+        intersection_generator.SkipDegreeTwoNodes(intersection_node, lhs.eid);
+    const auto right_connection =
+        intersection_generator.SkipDegreeTwoNodes(intersection_node, rhs.eid);
+
+    const auto left_candidate = node_based_graph.GetTarget(left_connection.via_eid);
+    const auto right_candidate = node_based_graph.GetTarget(right_connection.via_eid);
+
+    const auto candidate_is_valid =
+        left_candidate == right_candidate && left_candidate != intersection_node;
+
+    if (!candidate_is_valid)
+        return false;
+
+    // check if all entries at the destination or at the source are the same
+    const auto all_same_name_and_degree_three = [this](const NodeID nid) {
+        // check if the intersection found has degree three
+        if (node_based_graph.GetOutDegree(nid) != 3)
+            return false;
+
+        // check if all items share a name
+        const auto range = node_based_graph.GetAdjacentEdgeRange(nid);
+        const auto required_name_id = node_based_graph.GetEdgeData(range.front()).name_id;
+
+        const auto has_required_name = [this, required_name_id](const auto edge_id) {
+            return node_based_graph.GetEdgeData(edge_id).name_id == required_name_id;
+        };
+
+        /* the beautiful way would be:
+         * return range.end() == std::find_if_not(range.begin(), range.end(), has_required_name);
+         * but that does not work due to range concepts
+         */
```

Comment:
Wait, the iteration version should work, no? I thought only the boost::range version does not because of our range's limitations?